src/jit/flowgraph.cpp

```diff
@@ -2407,33 +2407,70 @@ BlockSet_ValRet_T   Compiler::fgDomFindStartNodes()
     return startNodes;
 }
 
-/** A simple DFS traversal of the flow graph.
-  * It computes both preorder and postorder numbering.
+/** A non-recursive DFS traversal of the flow graph using an
+  * evaluation stack to assign post-order numbers.
   */
 void Compiler::fgDfsInvPostOrderHelper(BasicBlock* block, BlockSet& visited, unsigned* count)
 {
     // Assume we haven't visited this node yet (callers ensure this).
     assert(!BlockSetOps::IsMember(this, visited, block->bbNum));
 
+    // Allocate a local stack to hold the Dfs traversal actions necessary
+    // to compute pre/post-ordering of the control flow-graph.
+    ArrayStack<DfsBlockEntry> stack(this);
+
+    // Push the first block on the stack to seed the traversal.
+    stack.Push(DfsBlockEntry(DSS_Pre, block));
     // Flag the node we just visited to avoid backtracking.
     BlockSetOps::AddElemD(this, visited, block->bbNum);
 
-    unsigned cSucc = block->NumSucc(this);
-    for (unsigned j = 0; j < cSucc; ++j)
+    // The search is terminated once all the actions have been processed.
+    while (stack.Height() != 0)
     {
-        BasicBlock* succ = block->GetSucc(j, this);
-        // If this is a node we haven't seen before, go ahead and recurse
-        if (!BlockSetOps::IsMember(this, visited, succ->bbNum))
+        DfsBlockEntry entry = stack.Pop();
```

Comment:
I would use different names, e.g., current and currentBlock, instead of entry and entryBlock here since there is just one entry block in a flow graph.